### PR TITLE
WooCommerce: Try the data-layer for product create

### PR DESCRIPTION
**Pull request by coderkevin**
This moves the http code to an action that is handled by the
extensions-data-layer middleware.

### CODE REVIEW

File: client/extensions/woocommerce/index.js [hunk 1]
```diff
@@ -178,3 +183,6 @@ export default function() {
 		page( '/store/stats/:type/:unit/:site', siteSelection, navigation, StatsController );
 	}
 }
+
+initExtension();
```
Comment (ockham) [inline client/extensions/woocommerce/index.js]: I find this a bit inconsistent with how we're automatically loading extension reducers. Could/Should we have `extensions-loader` take care of data-layer middleware, too?

Comment (ockham) [inline client/extensions/woocommerce/index.js]: cc @dmsnell ^

File: client/extensions/woocommerce/state/data-layer/products/index.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+ * Internal dependencies
+ */
+import { dispatchRequest } from 'state/data-layer/wpcom-http/utils';
+import { http } from 'state/data-layer/wpcom-http/actions';
+import { setError } from 'woocommerce/state/site/status/wc-api/actions';
+import {
+	WOOCOMMERCE_API_CREATE_PRODUCT,
+} from 'woocommerce/state/action-types';
+
+function createProduct( { dispatch }, action, next ) {
```
Comment (dmsnell) [inline client/extensions/woocommerce/state/data-layer/products/index.js]: Let's get rid of next here

File: client/extensions/woocommerce/state/data-layer/products/index.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+		method: 'POST',
+		query: httpProps,
+	}, action ) );
+
+	return next( action );
+}
+
+function createProductSuccess( { dispatch }, action, next, response ) {
+	const { siteId, successAction } = action.payload;
+	const { data } = response;
```
Comment (dmsnell) [inline client/extensions/woocommerce/state/data-layer/products/index.js]: Where is payload coming from? Was it there already?

Comment (coderkevin) [inline client/extensions/woocommerce/state/data-layer/products/index.js]: It was, but I'm removing it and putting properties directly in the action.

File: client/extensions/woocommerce/state/data-layer/products/index.js (hunk 2 above)
Comment (coderkevin) [inline client/extensions/woocommerce/state/data-layer/products/index.js]: Gotcha.

File: client/extensions/woocommerce/index.js (hunk 1 above)
Comment (coderkevin) [inline client/extensions/woocommerce/index.js]: I'm not sure. Maybe? I'll add a TODO to this line to remind us to revisit this approach. I'm definitely up for something better here.

File: client/extensions/woocommerce/state/sites/request.js [hunk 4]
```diff
@@ -31,6 +31,7 @@ const _request = ( method, path, siteId, body ) => {
 		{
 			path: `/wc/v3/${ path }&_method=${ method }`,
 			body: body && JSON.stringify( body ),
+			json: true,
```
Comment (coderkevin) [inline client/extensions/woocommerce/state/sites/request.js]: I found this to be necessary for the POST to work correctly. cc: @DanReyLop

Comment (DanReyLop) [inline client/extensions/woocommerce/state/sites/request.js]: Huh, really? It worked for me without `json: true`... that's weird... but it doesn't hurt.

Comment (coderkevin) [inline client/extensions/woocommerce/state/sites/request.js]: Yeah, I get this response:
`{"code":500,"headers":[{"name":"Content-Type","value":"application\/json"}],"body":{"error":"no_response_body","message":"Server could not read response."}}`

But adding `json: true` fixes it. However, it appears adding `json: true` has broken a few tests somehow. Not sure why that would be.

Comment (coderkevin) [inline client/extensions/woocommerce/state/sites/request.js]: Tests fixed, updating.

File: client/extensions/woocommerce/state/wc-api/products/reducer.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+ */
+import {
+	WOOCOMMERCE_PRODUCT_UPDATED,
+} from '../../action-types';
+
+export default {
+	[ WOOCOMMERCE_PRODUCT_UPDATED ]: productUpdated,
+};
+
+function productUpdated( siteData, action ) {
```
Comment (justinshreve) [inline client/extensions/woocommerce/state/wc-api/products/reducer.js]: Duplicate file probably left over from rebase.

File: client/extensions/woocommerce/state/data-layer/products/index.js [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+		return;
+	}
+
+	request( siteId ).post( 'products', productData )
+		.then( data => {
+			dispatch( {
+				type: WOOCOMMERCE_PRODUCT_CREATE_SUCCESS,
+				siteId,
+				data,
+				successAction,
```
Comment (belcherj) [inline client/extensions/woocommerce/state/data-layer/products/index.js]: are you missing an errorAction here?

File: client/extensions/woocommerce/state/wc-api/products/reducer.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+	const { data } = action.payload;
+
+	return updateCachedProduct( siteData, data );
+}
+
+function updateCachedProduct( siteData, product ) {
+	const products = siteData.products || [];
+
+	let found = false;
+	const newProducts = products.map( ( p ) => {
```
Comment (belcherj) [inline client/extensions/woocommerce/state/wc-api/products/reducer.js]: is it standard practice to use one letter variable names when doing maps or loops?

File: client/extensions/woocommerce/state/sites/products/actions.js [hunk 8]
```diff
@@ -1,91 +1,54 @@
...
- * API call to create a product within designated WooCommerce site.
+ * Action Creator: Create a new product
  *
- * @param {Number} siteId The Jetpack Site ID for the WooCommerce WordPress site.
- * @param {Object} product The product data to be created.
- * @param {Function} successAction Action creator, runs upon success, format of `function( data )`
- * @param {Function} failureAction Action creator, runs upon failure, format of `function( err )`
- * @returns {Object|Function} Object or thunk to be dispatched
+ * @param {Number} siteId The id of the site upon which to create the product.
+ * @param {Object} product The complete product object (may include a placeholder id)
```
Comment (coderkevin) [inline client/extensions/woocommerce/state/sites/products/actions.js]: It's theoretically possible. Maybe in a very simple bulk update or import?

File: client/extensions/woocommerce/state/data-layer/products/index.js (hunk 6 above)
Comment (coderkevin) [inline client/extensions/woocommerce/state/data-layer/products/index.js]: Good catch, thanks!

File: client/extensions/woocommerce/state/wc-api/products/reducer.js (hunk 7 above)
Comment (coderkevin) [inline client/extensions/woocommerce/state/wc-api/products/reducer.js]: I've seen it elsewhere. And I'm already using `product` as the incoming argument. Is this a problem?

File: client/extensions/woocommerce/state/wc-api/products/reducer.js (hunk 5 above)
Comment (coderkevin) [inline client/extensions/woocommerce/state/wc-api/products/reducer.js]: Thanks

**Comment by coderkevin** [comment]
@belcherj and @justinshreve, I have addressed your comments and put up changes for them. Please re-review. Thanks!

**Comment by coderkevin** [comment]
I just rebased this again on master and a test is failing. Will investigate and put up fix momentarily.

**Comment by justinshreve** [comment]
Took one last look - everything LGTM! All tests passing, and my product saves. :shipit:!

**Comment by coderkevin** [comment]
Thanks, Justin!

File: client/extensions/woocommerce/state/sites/products/reducer.js [hunk 9]
```diff
@@ -34,5 +32,6 @@ function createProductSuccess( state, action ) {
 		newProducts.push( product );
 	}
 
-	return state;
+	return { ...state, product: newProducts };
```
Comment (justinshreve) [inline client/extensions/woocommerce/state/sites/products/reducer.js]: @coderkevin Sorry for not catching this before, but this should be `products: newProducts`.

**Pull request merged**